Strip extra null characters in socketAddress

dlech opened the pull request:
Fixes <https://bugzilla.xamarin.com/show_bug.cgi?id=35004>.

Also remove unnecessary copying of array.

akoeplinger commented on mcs/class/Mono.Posix/Mono.Unix/UnixEndPoint.cs:
```diff
@@ -82,12 +82,10 @@ public override EndPoint Create (SocketAddress socketAddress)
-			}
 
-			string name = Encoding.Default.GetString (bytes);
+			string name = Encoding.Default.GetString (bytes, 2, socketAddress.Size - 2 - 1);
```
`bytes` doesn't exist here.

akoeplinger commented on mcs/class/Mono.Posix/Mono.Unix/UnixEndPoint.cs:
```diff
@@ -82,12 +82,10 @@ public override EndPoint Create (SocketAddress socketAddress)
-			string name = Encoding.Default.GetString (bytes);
+			string name = Encoding.Default.GetString (bytes, 2, socketAddress.Size - 2 - 1);
+			// remove any extra null characters
+			name = name.Replace ("\0", "");
```
this could be `name = name.TrimEnd ('\0');`

akoeplinger commented:
Can you please add the unit test to https://github.com/mono/mono/tree/master/mcs/class/Mono.Posix/Test/Mono.Unix? (we don't use the `Assert.That ()` NUnit style, so please adapt to the existing style).

dlech commented:
Thanks for the review. Yes, I will be glad to fix this up.

dlech commented:
Well, I added some unit tests, but it turns out that my fix doesn't work. I am getting two extra random characters when running the unit test instead of two extra null characters. I'll have to do some more digging to figure it out - unless someone else has any insight.

dlech commented:
OK. Should be good now. I've added a unit test for the original problem that I was seeing (UnixListenerTest) and a unit test for the method that was actually changed (UnixEndPointTest). Both of these tests fail on mono master. Then I added a separate commit with a new fix that is more robust than the previous one.

akoeplinger commented:
build

akoeplinger commented:
Looks good to me! I've triggered a build on the CI (expect a few test cases failing right now, they're unrelated).

Any other comments from people more familiar with the code?

dlech commented:
The build failed because I was using Path.GetFullPath() in one of the unit tests and the build server has a very long path name where the tests are running. Unix sockets are limited to 108 characters in the path (see `man 7 unix`). This is why it worked locally for me and failed in CI. We have to have an absolute path. I tried it and the test fails if I use a relative path. So, I have replaced Path.GetFullPath() with Path.GetTempFileName(), which should hopefully get a shorter file name in CI.  Please give it another go.

akoeplinger commented:
build

akoeplinger commented:
It looks good to me.

@jonpryor can you please take a look as well?

akoeplinger commented:
@dlech would you mind rebasing this to get rid of the merge conflicts?

dlech commented:
rebased

akoeplinger commented:
build

akoeplinger commented:
LGTM.

monojenkins commented:
Hello! I'm the build bot for the Mono project. I need approval from a Mono team member to build this pull request. A team member should reply with "approve" to approve a build of this pull request, "whitelist" to whitelist this and all future pull requests from this contributor, or "build" to explicitly request a build, even if one has already been done. Contributors can ignore this message.
